File: crates/better-core/src/sbom.rs
use std::collections::HashMap;
use std::fs;
use std::path::Path;

use crate::types::{SbomComponent, SbomReport};
use crate::{extract_json_field, resolve_from_lockfile, scan_licenses, JsonWriter};

// === D.6: SBOM export (CycloneDX + SPDX) ===

pub fn generate_sbom(project_root: &Path, lockfile: &Path, format: &str) -> Result<SbomReport, String> {
    let resolve_result = resolve_from_lockfile(lockfile)?;
    let nm = project_root.join("node_modules");
    let license_report = scan_licenses(&nm, &[], &[])?;
    let license_map: HashMap<String, String> = license_report.packages.iter()
        .map(|p| (p.name.clone(), p.license.clone())).collect();
    let mut components = Vec::new();
    for pkg in &resolve_result.packages {
        let license = license_map.get(&pkg.name).cloned().unwrap_or_else(|| "NOASSERTION".into());
        let purl = format!("pkg:npm/{}@{}", pkg.name, pkg.version);
        components.push(SbomComponent {
            name: pkg.name.clone(), version: pkg.version.clone(),
            license, purl, integrity: pkg.integrity.clone(),
        });
    }
    let pj = project_root.join("package.json");
    let c = fs::read_to_string(&pj).unwrap_or_default();
    let project_name = extract_json_field(&c, "name").unwrap_or_else(|| "unknown".into());
    let project_version = extract_json_field(&c, "version").unwrap_or_else(|| "0.0.0".into());
    Ok(SbomReport { format: format.into(), components, project_name, project_version })
}

pub fn write_cyclonedx_json(report: &SbomReport) -> String {
    let mut w = JsonWriter::new();
    w.begin_object();
    w.key("bomFormat"); w.value_string("CycloneDX");
    w.key("specVersion"); w.value_string("1.5");
    w.key("version"); w.value_i64(1);
    w.key("metadata"); w.begin_object();
    w.key("component"); w.begin_object();
    w.key("type"); w.value_string("application");
    w.key("name"); w.value_string(&report.project_name);
    w.key("version"); w.value_string(&report.project_version);
    w.end_object();
    w.end_object();
    w.key("components"); w.begin_array();
    for comp in &report.components {
        w.begin_object();
        w.key("type"); w.value_string("library");
        w.key("name"); w.value_string(&comp.name);
        w.key("version"); w.value_string(&comp.version);
        w.key("purl"); w.value_string(&comp.purl);
        w.key("licenses"); w.begin_array();
        w.begin_object();
        w.key("license"); w.begin_object();
        w.key("id"); w.value_string(&comp.license);
        w.end_object();
        w.end_object();
        w.end_array();
        w.end_object();
    }
    w.end_array();
    w.end_object();
    w.out.push('\n');
    w.finish()
}

pub fn write_spdx_json(report: &SbomReport) -> String {
    let mut w = JsonWriter::new();
    w.begin_object();
    w.key("spdxVersion"); w.value_string("SPDX-2.3");
    w.key("dataLicense"); w.value_string("CC0-1.0");
    w.key("SPDXID"); w.value_string("SPDXRef-DOCUMENT");
    w.key("name"); w.value_string(&report.project_name);
    w.key("documentNamespace"); w.value_string(
        &format!("https://spdx.org/spdxdocs/{}-{}", report.project_name, report.project_version)
    );
    w.key("packages"); w.begin_array();
    for (i, comp) in report.components.iter().enumerate() {
        w.begin_object();
        w.key("SPDXID"); w.value_string(&format!("SPDXRef-Package-{}", i));
        w.key("name"); w.value_string(&comp.name);
        w.key("versionInfo"); w.value_string(&comp.version);
        w.key("downloadLocation"); w.value_string(&comp.purl);
        w.key("licenseDeclared"); w.value_string(&comp.license);
        w.key("externalRefs"); w.begin_array();
        w.begin_object();
        w.key("referenceCategory"); w.value_string("PACKAGE-MANAGER");
        w.key("referenceType"); w.value_string("purl");
        w.key("referenceLocator"); w.value_string(&comp.purl);
        w.end_object();
        w.end_array();
        w.end_object();
    }
    w.end_array();
    w.end_object();
    w.out.push('\n');
    w.finish()
}

File: src/commands/sbom.js
import { parseArgs } from "node:util";
import { printJson, printText } from "../lib/output.js";
import { getRuntimeConfig } from "../lib/config.js";
import { findBetterCore } from "../lib/core.js";
import { runCommand } from "../lib/spawn.js";
import { join } from "node:path";
import fs from "node:fs/promises";

function spdxId(name, version) {
  return `SPDXRef-Package-${name.replace(/[^a-zA-Z0-9.-]/g, "-")}-${version.replace(/[^a-zA-Z0-9.-]/g, "-")}`;
}

function buildSpdx(pkgJson, components) {
  const now = new Date().toISOString();
  const docNamespace = `https://spdx.org/spdxdocs/${pkgJson.name || "unknown"}-${pkgJson.version || "0.0.0"}-${Date.now()}`;
  const lines = [
    "SPDXVersion: SPDX-2.3",
    "DataLicense: CC0-1.0",
    "SPDXID: SPDXRef-DOCUMENT",
    `DocumentName: ${pkgJson.name || "unknown"}`,
    `DocumentNamespace: ${docNamespace}`,
    `Creator: Tool: better-npm`,
    `Created: ${now}`,
    "",
    `PackageName: ${pkgJson.name || "unknown"}`,
    "SPDXID: SPDXRef-Package-root",
    `PackageVersion: ${pkgJson.version || "0.0.0"}`,
    "FilesAnalyzed: false",
    `PackageLicenseConcluded: ${pkgJson.license || "NOASSERTION"}`,
    `PackageLicenseDeclared: ${pkgJson.license || "NOASSERTION"}`,
    "PackageCopyrightText: NOASSERTION",
    ""
  ];
  for (const c of components) {
    lines.push(`PackageName: ${c.name}`);
    lines.push(`SPDXID: ${spdxId(c.name, c.version)}`);
    lines.push(`PackageVersion: ${c.version}`);
    lines.push(`PURL: ${c.purl}`);
    if (c.resolved) lines.push(`PackageDownloadLocation: ${c.resolved}`);
    else lines.push("PackageDownloadLocation: NOASSERTION");
    lines.push("FilesAnalyzed: false");
    lines.push(`PackageLicenseConcluded: ${c.license || "NOASSERTION"}`);
    lines.push(`PackageLicenseDeclared: ${c.license || "NOASSERTION"}`);
    lines.push("PackageCopyrightText: NOASSERTION");
    if (c.integrity) lines.push(`PackageChecksum: SHA-512: ${c.integrity.replace(/^sha512-/, "")}`);
    lines.push("");
  }
  for (const c of components) {
    lines.push(`Relationship: SPDXRef-Package-root DEPENDS_ON ${spdxId(c.name, c.version)}`);
  }
  return lines.join("\n") + "\n";
}

function buildCycloneDx(pkgJson, components, opts = {}) {
  const { includeVex = false } = opts;
  const sbom = {
    bomFormat: "CycloneDX",
    specVersion: "1.6",
    serialNumber: `urn:uuid:${Date.now().toString(16)}-${Math.random().toString(16).slice(2)}`,
    version: 1,
    metadata: {
      timestamp: new Date().toISOString(),
      tools: [{ vendor: "better-npm", name: "better", version: "0.3.0" }],
      component: {
        type: "application",
        bom_ref: "root",
        name: pkgJson.name || "unknown",
        version: pkgJson.version || "0.0.0",
        purl: `pkg:npm/${encodeURIComponent(pkgJson.name || "unknown")}@${pkgJson.version || "0.0.0"}`
      }
    },
    components: components.map(c => ({
      type: "library",
      "bom-ref": `${c.name}@${c.version}`,
      name: c.name,
      version: c.version,
      purl: c.purl,
      ...(c.resolved ? { externalReferences: [{ type: "distribution", url: c.resolved }] } : {}),
      ...(c.integrity ? { hashes: [{ alg: "SHA-512", content: c.integrity.replace(/^sha512-/, "") }] } : {}),
      ...(c.license ? { licenses: [{ license: { id: c.license } }] } : {})
    })),
    dependencies: [
      {
        ref: "root",
        dependsOn: components.map(c => `${c.name}@${c.version}`)
      }
    ]
  };
  if (includeVex) {
    sbom.vulnerabilities = [];
  }
  return sbom;
}

export async function cmdSbom(argv) {
  if (argv.includes("--help") || argv.includes("-h")) {
    printText(`Usage: better sbom [options]

Generate a Software Bill of Materials (SBOM) for installed packages.

Options:
  --format <fmt>    Output format: cyclonedx (default) | spdx
  --output <file>   Write to file instead of stdout
  --vex             Include VEX (Vulnerability Exploitability eXchange) section
  --project-root    Path to project root
  --json            Machine-readable output wrapper
  -h, --help        Show this help

Examples:
  better sbom                                 # CycloneDX 1.6 JSON to stdout
  better sbom --format spdx --output sbom.spdx
  better sbom --vex --output bom.json
`);
    return;
  }

  const runtime = getRuntimeConfig();
  const { values } = parseArgs({
    args: argv,
    options: {
      json: { type: "boolean", default: runtime.json === true },
      format: { type: "string", default: "cyclonedx" },
      output: { type: "string" },
      vex: { type: "boolean", default: false },
      "project-root": { type: "string" }
    },
    allowPositionals: false,
    strict: false
  });

  const useJson = values.json;
  const format = String(values.format ?? "cyclonedx").toLowerCase();
  if (format !== "cyclonedx" && format !== "spdx") {
    printText(`Unknown format '${format}'. Expected cyclonedx|spdx.`);
    process.exitCode = 1;
    return;
  }

  const cwd = values["project-root"] ? String(values["project-root"]) : process.cwd();

  // Try to delegate to better-core binary
  const binPath = await findBetterCore();
  if (binPath) {
    const args = ["sbom"];
    if (values.format) args.push("--format", values.format);
    if (values.output) args.push("--output", values.output);
    if (values.vex) args.push("--vex");
    if (useJson) args.push("--json");
    const result = await runCommand(binPath, args, { cwd, passthroughStdio: true });
    process.exitCode = result.exitCode ?? 0;
    return;
  }

  // JS fallback: generate SBOM from package-lock.json
  try {
    const lockPath = join(cwd, "package-lock.json");
    const lock = JSON.parse(await fs.readFile(lockPath, "utf8"));
    const pkgJson = JSON.parse(await fs.readFile(join(cwd, "package.json"), "utf8"));
    const packages = lock.packages ?? {};

    const components = [];
    for (const [pkgPath, info] of Object.entries(packages)) {
      if (!pkgPath || pkgPath === "") continue;
      const name = pkgPath.startsWith("node_modules/")
        ? pkgPath.slice("node_modules/".length)
        : pkgPath;
      if (!name || name.includes("/node_modules/")) continue;
      components.push({
        name,
        version: info.version || "0.0.0",
        purl: `pkg:npm/${encodeURIComponent(name)}@${info.version || "0.0.0"}`,
        integrity: info.integrity ?? null,
        resolved: info.resolved ?? null,
        license: info.license ?? null
      });
    }

    if (format === "spdx") {
      const spdxText = buildSpdx(pkgJson, components);
      if (values.output) {
        await fs.writeFile(values.output, spdxText);
        if (!useJson) printText(`SBOM (SPDX 2.3) written to ${values.output} (${components.length} packages)`);
        else printJson({ ok: true, kind: "better.sbom", format: "spdx", specVersion: "2.3", components: components.length, output: values.output });
      } else {
        process.stdout.write(spdxText);
      }
      return;
    }

    // CycloneDX 1.6
    const sbom = buildCycloneDx(pkgJson, components, { includeVex: values.vex === true });
    const output = JSON.stringify(sbom, null, 2);

    if (values.output) {
      await fs.writeFile(values.output, output + "\n");
      if (!useJson) printText(`SBOM (CycloneDX 1.6) written to ${values.output} (${components.length} components)`);
      else printJson({ ok: true, kind: "better.sbom", format: "cyclonedx", specVersion: "1.6", components: components.length, output: values.output });
    } else if (useJson) {
      printJson({ ok: true, kind: "better.sbom", format: "cyclonedx", specVersion: "1.6", ...sbom });
    } else {
      process.stdout.write(output + "\n");
    }
  } catch (err) {
    if (useJson) printJson({ ok: false, kind: "better.sbom", error: err.message });
    else printText(`Error: ${err.message}`);
    process.exitCode = 1;
  }
}
